### PR TITLE
Typespec renaming

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/3196
Fix https://github.com/Azure/autorest.csharp/issues/3183

- upgrade typespec libraries to 0.41.0
- rename dotnet emitter to @azure-tools/typespec-csharp 

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "private": true,
   "description": "package.json intended for in-repo use only, package.json used for publishing is located in src/AutoRest.CSharp/package.json",
   "devDependencies": {
-    "@azure-tools/cadl-csharp": "src/CADL.Extension/Emitter.Csharp",
+    "@azure-tools/typespec-csharp": "src/CADL.Extension/Emitter.Csharp",
     "@azure-tools/cadl-ranch-mockapis": "file:test/CadlRanchMockApis",
```
Comment (pshao25) [inline package.json:7]: Could you rename cadl-ranch-mockapis as well?

File: src/CADL.Extension/Emitter.Csharp/package.json [hunk 2]
```diff
@@ -1,10 +1,10 @@
 {
-    "name": "@azure-tools/cadl-csharp",
+    "name": "@azure-tools/typespec-csharp",
     "version": "0.1.11",
```
Comment (pshao25) [inline src/CADL.Extension/Emitter.Csharp/package.json:3]: We should bump emitter version to release a new one.

Comment (pshao25) [inline src/CADL.Extension/Emitter.Csharp/package.json:3]: We should bump emitter version to release a new one.

Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/package.json:3]: Yes. I prefer to upgrade to 0.2.0

File: package.json (hunk 1 above)
Comment (chunyu3) [reply package.json:7]: cadl-ranch have not updated the name. For consistent, we will delay to update this until cadl-ranch updates the name.  And will update all the test project folder name to remove `cadl`, such as `cadlRanchProjects`, `cadlRanchMockApis` and so on.

**Review by lirenhe — COMMENTED** [review]
Do you plan to rename files also in this PR?  
CadlGenerate.ps1 =>typespecGenerate.ps1

File: eng/CadlGenerate.ps1 [hunk 3]
```diff
@@ -47,7 +47,7 @@ foreach ($directory in Get-ChildItem $samplePath -Directory)
      {
          Remove-Item $projectPath/Generated -Force -Recurse
      }
-    node node_modules/@cadl-lang/compiler/dist/core/cli.js compile $projectPath/$testName.cadl --emit @azure-tools/cadl-csharp --option @azure-tools/cadl-csharp.emitter-output-dir=$projectPath --option @azure-tools/cadl-csharp.skipSDKGeneration=true --option @azure-tools/cadl-csharp.save-inputs=true
+    node node_modules/@typespec/compiler/dist/core/cli/cli.js compile $projectPath/$testName.tsp --emit @azure-tools/typespec-csharp --option @azure-tools/typespec-csharp.emitter-output-dir=$projectPath --option @azure-tools/typespec-csharp.skipSDKGeneration=true --option @azure-tools/typespec-csharp.save-inputs=true
     if (!$?) {
         Pop-Location
         throw "Failed to emit cadl model for $testName."
```
Comment (lirenhe) [inline eng/CadlGenerate.ps1:53]:
```suggestion
        throw "Failed to emit typespec model for $testName."
```

File: eng/Generate.ps1 [hunk 4]
```diff
@@ -53,11 +53,11 @@ function Add-TestServer-Swagger ([string]$testName, [string]$projectSuffix, [str
     Add-Swagger "$testName$projectSuffix" $projectDirectory "--require=$configurationPath --try-require=$inputReadme --input-file=$inputFile $additionalArgs"
 }
 
-function Add-CadlRanch-Cadl([string]$testName, [string]$projectPrefix, [string]$cadlRanchProjectsDirectory, [boolean]$generateConvenience) {
+function Add-CadlRanch-Typespec([string]$testName, [string]$projectPrefix, [string]$cadlRanchProjectsDirectory, [boolean]$generateConvenience) {
     $projectDirectory = Join-Path $cadlRanchProjectsDirectory $testName
     $cadlMain = Join-Path $cadlRanchFilePath $testName "main.cadl"
```
Comment (lirenhe) [inline eng/Generate.ps1:58]: Do we still need to use "main.cadl"

Comment (chunyu3) [reply eng/Generate.ps1:58]: Yes. cadl-ranch has not updated the file name to *.tsp yet. So we still need main.cadl.

**Review by pshao25 — APPROVED** [review]

**Comment by lirenhe** [comment]
Could you share the result of the SDK repo regen?  
Any other failures besides OpenAI and MVAD?

**Comment by chunyu3** [comment]
> @chunyu3 Could you share the result of the SDK repo regen? Any other failures besides OpenAI and MVAD?

Yes. I regened the sdk repo. There is no changes except failure on cadl dpg project OpenAI and MVAD.

**Comment by chunyu3** [comment]
regen azure-sdk-for-net, regen branch: https://github.com/chunyu3/azure-sdk-for-net/tree/typespecRenaming, no sdk change.

File: src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts [hunk 5]
```diff
@@ -39,7 +39,7 @@ function getDefaultValue(type: Type): any {
 
 export function resolveServers(
     program: Program,
-    servers: http.HttpServer[],
+    servers: HttpServer[],
     models: Map<string, InputModelType>,
     enums: Map<string, InputEnumType>
 ): CadlServer[] {
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts:45]: Do we still need 'CadlServer' ? 
And is it possible to rename this file as well?

File: src/CADL.Extension/Emitter.Csharp/package.json [hunk 6]
```diff
@@ -1,10 +1,10 @@
 {
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/package.json:1]: Do you plan to do the folder rename of 'CADL.Extension' in another PR?

File: src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts (hunk 5 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts:45]: We have not update the source code. we will update in another PR together with the project folder name

File: src/CADL.Extension/Emitter.Csharp/samples/CadlFirstTest/CadlFirstTest.tsp [hunk 7]
```diff
@@ -1,5 +1,6 @@
-import "@cadl-lang/rest";
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/samples/CadlFirstTest/CadlFirstTest.tsp:1]: Same above for those file/folder rename.

File: src/CADL.Extension/Emitter.Csharp/test/TestProjects/enum/Generated/cadl.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/test/TestProjects/enum/Generated/cadl.json:1]: Will you consider use another name instead of cadl.json?

**Comment by lirenhe** [comment]
The rest of rename (file/folder, cadl.json etc )could be happen in another PR.  Approve this one to unblock the migration effort.

**Review by lirenhe — APPROVED** [review]

**Pull request merged**